src/Streamly/Coreutils/Cp.hs

```diff
@@ -0,0 +1,281 @@
+removeTrailingSlash "" = ""
+removeTrailingSlash path =
+    let end = last path
+        osName = os
+        pathSep = if osName == "windows" then '\\' else '/'
+     in
+         if end == pathSep
+         then init path
+         else path
+
+
+-- | Lists the contents (both files and directories) of the directory
+-- relative to the argument directory.
+-- @Left path@ represents that @path@ is a directory while @Right path@
+-- represents that it is a file.
+--
+-- @since 0.1.0.0
+traverseDir
+    :: IsStream t
+    => FilePath
+    -> t IO (Either FilePath FilePath)
+traverseDir baseDir = do
+    IP.concatMapTreeWith ahead listContents $ S.yield $ Left ""
+
+    where
+
+    listContents :: IsStream t => FilePath -> t IO (Either FilePath FilePath)
+    listContents dir = do
+        S.fromListM $ map (identify dir) $ contents dir
+
+    identify :: FilePath -> FilePath -> IO (Either FilePath FilePath)
+    identify dir path = do
+       let actualPath = append baseDir $ append dir path
+       isFile <- doesFileExist actualPath
+       if isFile == True
+       then
+           return $ Right $ append dir path     -- RELATIVE File
+       else
+           return $ Left $ append dir path      -- RELATIVE Dir
+
+    contents :: FilePath -> [FilePath]
+    contents dir = unsafePerformIO $ listDirectory $ append baseDir dir
+
+
+
+-- | Recursively copies the contents of the source directory
+-- to the destination directory.
+-- Arguments are of type @FilePath@.
+--
+-- @since 0.1.0.0
+copyDirToDir
+    :: IsStream t
+    => CpOptions
+    -> FilePath
+    -> FilePath
+    -> t IO ()
+copyDirToDir opt src dest = do
+    if force opt == True
+    then
+        S.mapM handleForce
+        $ traverseDir src
+    else
+        S.mapM (\_ -> return ())
+        $ S.takeWhileM handleNonForce
+        $ traverseDir src
+
+    where
+
+    handleForce :: Either FilePath FilePath -> IO ()
+    handleForce (Left dir) = createDirectoryIfMissing True $ append dest dir
+    handleForce (Right file) =
+        ( copy opt (append src file) (append dest file)
+        >> return ()
+        )
+
+    handleNonForce :: Either FilePath FilePath -> IO Bool
+    handleNonForce (Left dir) =
+        ( createDirectoryIfMissing True (append dest dir)
+        >> return True
+        )
+    handleNonForce (Right file) =
+        copy opt (append src file) (append dest file)
+
+
+-- | Copies the source file to the
+-- destination path with a possible rename.
+--
+-- @since 0.1.0.0
+cpFileWithRename
+    :: CpOptions
+    -> SomeBase File
+    -> SomeBase File
+    -> IO Bool
+cpFileWithRename opt src dst = do
+    let srcFp = removeTrailingSlash $ someFileToFP src
+    let dstFp = removeTrailingSlash $ someFileToFP dst
+    copy opt srcFp dstFp
+
+
+-- | Copies the source file preserving the source file name
+-- in the destination directory.
+-- Returns @True@ in the @IO@ monad if copying was successful,
+-- @False@ otherwise.
+--
+-- @since 0.1.0.0
+cpFile
```

Comment:
Why do you have this? Why not just use `Path Abs File` in `copy`?

Reply:
This is the `File -> Dir` case.